File: frontend/src/shared/api/academic.ts
import type { ClassListResponse, CourseListResponse, StudentListResponse } from "../types/academic";

const apiBaseUrl = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000/api";

async function requestJson<T>(path: string): Promise<T> {
  const response = await fetch(`${apiBaseUrl}${path}`);

  if (!response.ok) {
    throw new Error(`Request failed: ${response.status}`);
  }

  return response.json() as Promise<T>;
}

export function fetchCourses(): Promise<CourseListResponse> {
  return requestJson<CourseListResponse>("/courses");
}

export function fetchClasses(courseId = "course_web_2026"): Promise<ClassListResponse> {
  return requestJson<ClassListResponse>(`/courses/${courseId}/classes`);
}

export function fetchStudents(classId = "class_cs_2024_01"): Promise<StudentListResponse> {
  return requestJson<StudentListResponse>(`/classes/${classId}/students`);
}

File: backend/app/api/routes/academic.py
from fastapi import APIRouter

from app.schemas.academic import ClassListResponse, CourseListResponse, StudentListResponse
from app.services.academic_service import AcademicService

router = APIRouter()


@router.get("/courses", response_model=CourseListResponse)
def list_courses() -> CourseListResponse:
    return AcademicService().list_courses()


@router.get("/courses/{course_id}/classes", response_model=ClassListResponse)
def list_classes(course_id: str) -> ClassListResponse:
    return AcademicService().list_classes(course_id)


@router.get("/classes/{class_id}/students", response_model=StudentListResponse)
def list_students(class_id: str) -> StudentListResponse:
    return AcademicService().list_students(class_id)
